walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  
+  private Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    timeOptions.clear();
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    Set<String> allEventAttendees = new HashSet<>();
+    for (Event event : events) {
+      allEventAttendees.addAll(event.getAttendees());
+    }
+    Set<String> allRequestAttendees = new HashSet<>();
+    allRequestAttendees.addAll(request.getAttendees());
+    for (String attendee : allRequestAttendees) {
+      if (!allEventAttendees.contains(attendee)) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+    
+    List<Event> eventsList = new ArrayList<Event>();
+    eventsList.addAll(events);
+    Collections.sort(eventsList, Event.ORDER_BY_START);
+    TimeRange firstEventTime = eventsList.get(0).getWhen();
+    TimeRange lastEventTime = eventsList.get(eventsList.size() - 1).getWhen();
+    boolean earlyEnd = false;
+    if (TimeRange.START_OF_DAY != firstEventTime.start()) {
+      addTimeRange(TimeRange.START_OF_DAY, firstEventTime.start());
+    }
+    for (int i = 1; i < eventsList.size(); i++) {
+      TimeRange currentEventTime = eventsList.get(i - 1).getWhen();
+      TimeRange nextEventTime = eventsList.get(i).getWhen();
+      TimeRange skipNextTime = null; 
+      if (i + 1 < eventsList.size()) {
+        skipNextTime = eventsList.get(i + 1).getWhen();
+      }
+      if (!backToBack(currentEventTime, nextEventTime)) {
```

Comment:
Invert this conditional and use continue to avoid the nesting.

```java
    for (int i = 1; i < eventsList.size(); i++) {
      TimeRange currentEventTime = eventsList.get(i - 1).getWhen();
      TimeRange nextEventTime = eventsList.get(i).getWhen();
      TimeRange skipNextTime = null; 
      if (i + 1 < eventsList.size()) {
        skipNextTime = eventsList.get(i + 1).getWhen();
      }
      if (backToBack(currentEventTime, nextEventTime)) {
        // TODO: Explain, please.
        continue
      }
      if (!currentEventTime.overlaps(nextEventTime)) {
        addTimeRange(currentEventTime.end(), nextEventTime.start());
      }
      // (...)
    }
```

Reply:
Done